packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,10 +11,707 @@ import 'picker.dart';
+      remainingWidth -= columnWidth[i] + _kDatePickerPadSize * 2;
+
+    double currentHorizontalOffset = 0.0;
+
+    for (int i = 0; i < columnWidth.length; i++) {
+      final int index = textDirectionFactor == 1 ? i : columnWidth.length - i - 1;
+
+      double childWidth = columnWidth[index] + _kDatePickerPadSize * 2;
+      if (index == 0 || index == columnWidth.length - 1)
+        childWidth += remainingWidth / 2;
+
+      layoutChild(index, BoxConstraints.tight(new Size(childWidth, size.height)));
+      positionChild(index, new Offset(currentHorizontalOffset, 0.0));
+
+      currentHorizontalOffset += childWidth;
+    }
+  }
+
+  @override
+  bool shouldRelayout(_DatePickerLayout oldDelegate) {
+    return columnWidth != oldDelegate.columnWidth
+      || textDirectionFactor != oldDelegate.textDirectionFactor;
+  }
+}
+
+/// Different modes of [CupertinoDatePicker].
+enum CupertinoDatePickerMode {
+  /// Mode that shows the date in hour, minute, and (optional) an AM/PM designation.
+  ///
+  /// Example: [4 | 14 | PM].
+  time,
+  /// Mode that shows the date in month, day of month, and year.
+  ///
+  /// Example: [July | 13 | 2012].
+  date,
+  /// Mode that shows the date as day of the week, month, day of month and
+  /// the time in hour, minute, and (optional) an AM/PM designation.
+  ///
+  /// Example: [Fri Jul 13 | 4 | 14 | PM]
+  dateAndTime,
+}
+
+/// A date picker widget in iOS style.
+///
+/// There are several modes of the date picker listed in [CupertinoDatePickerMode].
+class CupertinoDatePicker extends StatefulWidget {
+  /// Constructs an iOS style date picker.
+  ///
+  /// [mode] is one of the mode listed in [CupertinoDatePickerMode] and defaults
+  /// to [CupertinoDatePickerMode.dateAndTime].
+  ///
+  /// [onDateTimeChanged] is the callback when the selected date or time changes
+  /// and must not be null.
+  ///
+  /// [initialDateTime] is the initial value of the picker. Must not be null.
+  ///
+  /// [minimumDate] is the minimum date that the picker can be scrolled to in
+  /// [CupertinoDatePickerMode.dateAndTime] mode. Null if there's no limit.
+  ///
+  /// [maximumDate] is the maximum date that the picker can be scrolled to in
+  /// [CupertinoDatePickerMode.dateAndTime] mode. Null if there's no limit.
+  ///
+  /// [minimumYear] is the minimum year that the picker can be scrolled to in
+  /// [CupertinoDatePickerMode.date] mode. Defaults to 1 and must not be null.
+  ///
+  /// [maximumYear] is the maximum year that the picker can be scrolled to in
+  /// [CupertinoDatePickerMode.date] mode. Null if there's no limit.
+  ///
+  /// [minuteInterval] is the granularity of the minute spinner. Must be a
+  /// positive integer factor of 60.
+  ///
+  /// [use24hFormat] decides whether 24 hour format is used. Defaults to false.
+  CupertinoDatePicker({
+    this.mode = CupertinoDatePickerMode.dateAndTime,
+    @required this.onDateTimeChanged,
+    @required this.initialDateTime,
```

Comment:
I can't set it to DateTime.now(), because default value has to be constant.
Setting it to some random value might conflict with date limit, so I think it's best for user to decide what the default value is.

Reply:
You can set it in the initializers before the asserts. You can document that now needs to be in the acceptable dates (which you already assert) or the user needs to provide a custom one.